kernel/reduction.ml

```diff
@@ -743,21 +788,38 @@ and convert_vect l2r infos lft1 lft2 v1 v2 cuniv =
     fold 0 cuniv
   else raise NotConvertible
 
-and convert_branches l2r infos ci e1 e2 lft1 lft2 br1 br2 cuniv =
-  (** Skip comparison of the pattern types. We know that the two terms are
-      living in a common type, thus this check is useless. *)
-  let fold n c1 c2 cuniv = match skip_pattern infos n c1 c2 with
-  | (infos, c1, c2) ->
-    let lft1 = el_liftn n lft1 in
-    let lft2 = el_liftn n lft2 in
+and convert_under_context l2r infos e1 e2 lft1 lft2 ctx u1 u2 args1 args2 (nas1, c1) (nas2, c2) cu =
+  let n = Array.length nas1 in
+  let () = assert (Int.equal n (Array.length nas2)) in
+  let n, e1, e2 = match ctx with
+  | None -> (* nolet *)
     let e1 = subs_liftn n e1 in
     let e2 = subs_liftn n e2 in
-    ccnv CONV l2r infos lft1 lft2 (mk_clos e1 c1) (mk_clos e2 c2) cuniv
-  | exception IrregularPatternShape ->
-    (** Might happen due to a shape invariant that is not enforced *)
-    ccnv CONV l2r infos lft1 lft2 (mk_clos e1 c1) (mk_clos e2 c2) cuniv
+    (n, e1, e2)
+  | Some ctx ->
+    let ctx1 = Vars.subst_instance_context u1 ctx in
+    let ctx2 = Vars.subst_instance_context u2 ctx in
+    let n1, e1 = esubst_of_rel_context_instance ctx1 args1 e1 in
+    let n2, e2 = esubst_of_rel_context_instance ctx2 args2 e2 in
+    let () = assert (Int.equal n1 n2) in
+    n1, e1, e2
+  in
+  let lft1 = el_liftn n lft1 in
+  let lft2 = el_liftn n lft2 in
+  let infos = push_relevances infos nas1 in
```

Comment:
So it's essentially for this line that we keep the relevances ?

Reply:
I think so. The compuation of relevance itself in `Relevanceops` also relies on that in order not to to access the env.

Reply:
Ok, we'll do it that way too in MetaCoq.